File: src/main/java/com/viaversion/aas/codec/packet/handshake/Handshake.java
package com.viaversion.aas.codec.packet.handshake;

import com.viaversion.aas.codec.packet.Packet;
import com.viaversion.aas.util.IntendedState;
import com.viaversion.viaversion.api.protocol.version.ProtocolVersion;
import com.viaversion.viaversion.api.type.Type;
import com.viaversion.viaversion.api.type.Types;
import io.netty.buffer.ByteBuf;
import org.jetbrains.annotations.NotNull;

public class Handshake implements Packet {
	private int protocolId;
	private String address;
	private int port;
	private IntendedState intendedState;

	@Override
	public void decode(@NotNull ByteBuf byteBuf, ProtocolVersion protocolVersion) throws Exception {
		protocolId = Types.VAR_INT.readPrimitive(byteBuf);
		address = Types.STRING.read(byteBuf);
		port = byteBuf.readUnsignedShort();
		intendedState = IntendedState.values()[Types.VAR_INT.readPrimitive(byteBuf)];
	}

	@Override
	public void encode(@NotNull ByteBuf byteBuf, ProtocolVersion protocolVersion) throws Exception {
		Types.VAR_INT.writePrimitive(byteBuf, protocolId);
		Types.STRING.write(byteBuf, address);
		byteBuf.writeShort(port);
		Types.VAR_INT.writePrimitive(byteBuf, intendedState.ordinal());
	}

	public int getProtocolId() {
		return protocolId;
	}

	public void setProtocolId(int protocolId) {
		this.protocolId = protocolId;
	}

	public String getAddress() {
		return address;
	}

	public void setAddress(String address) {
		this.address = address;
	}

	public int getPort() {
		return port;
	}

	public void setPort(int port) {
		this.port = port;
	}

	public IntendedState getIntendedState() {
		return intendedState;
	}

	public void setIntendedState(IntendedState intendedState) {
		this.intendedState = intendedState;
	}
}

File: src/main/kotlin/com/viaversion/aas/codec/packet/handshake/Handshake.kt
package com.viaversion.aas.codec.packet.handshake

import com.viaversion.aas.codec.packet.Packet
import com.viaversion.aas.config.VIAaaSConfig
import com.viaversion.viaversion.api.protocol.packet.State
import com.viaversion.viaversion.api.type.Type
import com.viaversion.viaversion.api.type.types.StringType
import io.netty.buffer.ByteBuf
import kotlin.properties.Delegates

class Handshake : Packet {
    var protocolId by Delegates.notNull<Int>()
    lateinit var address: String
    var port by Delegates.notNull<Int>()
    lateinit var nextState: State

    override fun decode(byteBuf: ByteBuf, protocolVersion: Int) {
        protocolId = Type.VAR_INT.readPrimitive(byteBuf)
        address = StringType((if (VIAaaSConfig.bungeeCord) Short.MAX_VALUE else 255).toInt()).read(byteBuf)
        port = byteBuf.readUnsignedShort()
        nextState = State.values()[Type.VAR_INT.readPrimitive(byteBuf)]
    }

    override fun encode(byteBuf: ByteBuf, protocolVersion: Int) {
        Type.VAR_INT.writePrimitive(byteBuf, protocolId)
        Type.STRING.write(byteBuf, address)
        byteBuf.writeShort(port)
        byteBuf.writeByte(nextState.ordinal) // var int is too small, fits in a byte
    }
}